Mathlib/CategoryTheory/Sites/PreservesSheafification.lean

```diff
@@ -0,0 +1,157 @@
+lemma sheafifyCompose_fac :
+    toSheafify J (P ⋙ F) ≫ sheafifyCompose J F P = whiskerRight (toSheafify J P) F := by
+  dsimp only [sheafifyCompose, toSheafify, presheafToSheafCompose]
+  erw [Adjunction.homEquiv_counit, Adjunction.unit_naturality_assoc]
+  simp
+
+/-- Given a Grothendieck topology `J` on `C` and `F : A ⥤ B`, this is the condition
+that sheafification for `J` commutes with the postcomposition with `F`. -/
+class PreservesSheafification : Prop where
+  isIso : IsIso (presheafToSheafCompose J F) := by infer_instance
+
+lemma PreservesSheafification.mk' (h : ∀ (P : Cᵒᵖ ⥤ A), IsIso (sheafifyCompose J F P)) :
+    PreservesSheafification J F where
+  isIso := by
+    have : ∀ P, IsIso ((sheafToPresheaf J B).map ((presheafToSheafCompose J F).app P)) := h
+    have : ∀ P, IsIso ((presheafToSheafCompose J F).app P) :=
+      fun p => isIso_of_reflects_iso _ (sheafToPresheaf J B)
+    apply NatIso.isIso_of_isIso_app
+
+variable [PreservesSheafification J F]
+
+attribute [instance] PreservesSheafification.isIso
+
+/-- Given a Grothendieck topology `J` on `C` and `F : A ⥤ B`
+such that `[PreservesSheafification J F]`, this is the condition
+that sheafification for `J` commutes with the postcomposition with `F`. -/
+noncomputable def presheafToSheafComposeIso :
+    (whiskeringRight Cᵒᵖ A B).obj F ⋙ presheafToSheaf J B ≅
+      presheafToSheaf J A ⋙ sheafCompose J F :=
+  asIso (presheafToSheafCompose J F)
+
+/-- The canonical isomorphism `sheafify J (P ⋙ F) ≅ sheafify J P ⋙ F` when
+`F` preserves the sheafification. -/
+noncomputable def sheafifyComposeIso :
+    sheafify J (P ⋙ F) ≅ sheafify J P ⋙ F :=
+  (sheafToPresheaf J B).mapIso ((presheafToSheafComposeIso J F).app P)
+
+@[simp]
+lemma sheafifyComposeIso_hom :
+    (sheafifyComposeIso J F P).hom = sheafifyCompose J F P := rfl
+
+@[reassoc (attr := simp)]
+lemma sheafifyComposeIso_hom_inv_id :
+    sheafifyCompose J F P ≫ (sheafifyComposeIso J F P).inv = 𝟙 _ :=
+  (sheafifyComposeIso J F P).hom_inv_id
+
+@[reassoc (attr := simp)]
+lemma sheafifyComposeIso_inv_hom_id :
+    (sheafifyComposeIso J F P).inv ≫ sheafifyCompose J F P = 𝟙 _ :=
+  (sheafifyComposeIso J F P).inv_hom_id
+
+instance : IsIso (sheafifyCompose J F P) :=
+  (inferInstance : IsIso (sheafifyComposeIso J F P).hom)
+
+@[reassoc (attr := simp)]
+lemma sheafifyComposeIso_inv_fac :
+    whiskerRight (toSheafify J P) F ≫ (sheafifyComposeIso J F P).inv =
+      toSheafify J (P ⋙ F) := by
+  rw [← cancel_mono (sheafifyCompose J F P), assoc, sheafifyComposeIso_inv_hom_id,
+    comp_id, sheafifyCompose_fac]
+
+end
+
+namespace GrothendieckTopology
+
+variable {D : Type*} [Category.{max v u} D]
+  [ConcreteCategory.{max v u} D] [PreservesLimits (forget D)]
+  [∀ X : C, HasColimitsOfShape (J.Cover X)ᵒᵖ D]
+  [∀ X : C, PreservesColimitsOfShape (J.Cover X)ᵒᵖ (forget D)]
+  [∀ (α β : Type max u v) (fst snd : β → α),
+      Limits.HasLimitsOfShape (Limits.WalkingMulticospan fst snd) D]
+  [(forget D).ReflectsIsomorphisms]
```

Comment:
Is it possible to prove this for general functors other than forget (or maybe at least forget_2) for this?

Reply:
Yes. Actually, I think I should generalize `PreservesSheafification` without even assuming `HasSheafCompose`. For example, in some applications, we would like to apply this to the free abelian group functor from sets to abelian groups (or similarly the free module functor). The postcomposition with this free abelian group functor should not preserve sheaves, but still it behaves quite well with respect to sheafification, in the sense that if we introduce the class `W` of morphisms of presheaves which become isos after sheafification, then `W` is "preserved" by the application of this free abelian group functor. If we assume moreover `HasSheafCompose`, then this condition should be equivalent to the current draft `PreservesSheafification`. (Eventually, for good enough concrete categories, the class `W` should coincide with the class of morphisms of presheaves that are both locally injective and locally surjective.)

Reply:
The class `W` is defined in #12374

Reply:
I have generalized the definition of `PreservesSheafification` as I suggested above.